test/test_linalg.py

```diff
@@ -121,6 +121,157 @@ def run_test_case(a, b):
         run_test_case(zero_strided, b)
         run_test_case(a, zero_strided)
 
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_linalg_lstsq(self, device, dtype):
+        from torch.testing._internal.common_utils import random_well_conditioned_matrix
+        drivers = ('gels', 'gelsy', 'gelsd', 'gelss', None) \
+            if device == 'cpu' \
+            else ('gels', None)
+
+        def check_correctness(a, b, sol):
+            sol2 = a.pinverse() @ b
+            self.assertEqual(sol, sol2, atol=1e-5, rtol=1e-5)
+
+        def check_ranks(a, ranks, cond=1e-7):
+            ranks2 = torch.matrix_rank(a, tol=cond)
+            ranks2 = ranks2.unsqueeze(0) if ranks2.dim() == 0 else ranks2
+            self.assertEqual(ranks, ranks2)
+
+        def check_singular_values(a, sv):
+            sv2 = a.svd()[1]
+            self.assertEqual(sv, sv2)
+
+        ms = [2 ** i for i in range(1, 5)]
+        m_ge_n_sizes = [(m, m // 2) for m in ms] + [(m, m) for m in ms]
+        m_l_n_sizes = [(m // 2, m) for m in ms]
+        matrix_sizes = m_ge_n_sizes + (m_l_n_sizes if device == 'cpu' else [])
+        batches = [(), (2,), (2, 2), (2, 2, 2)]
+
+        for batch, matrix_size, driver in itertools.product(batches, matrix_sizes, drivers):
+            shape = batch + matrix_size
+            a = random_well_conditioned_matrix(*shape, dtype=dtype, device=device)
+            b = torch.rand(*shape, dtype=dtype, device=device)
+
+            cond = 1e-7
+            m = a.size(-2)
+            n = a.size(-1)
+            res = torch.linalg.lstsq(a, b, cond=cond, driver_name=driver)
+            sol = res.solution.narrow(-2, 0, n)
+
+            check_correctness(a, b, sol)
+            if device == 'cpu':
+                if driver != 'gels':
```

Comment:
Style nit:

```
if self.device_type == 'cpu' and driver != 'gels':
  # comment explaining the conditional here
  check_ranks(a, res.rank, cond)
if self.device_type == 'cpu' and driver not in (None, 'gels', 'gelsy'):
  # comment explaining the conditional here
  check_singular_values(a, res.singular_values)
```

I think this will be more readable.